Add Continuous Flux Injection from Differential MCEq Profile as a new Body Class and Corresponding Updates

ConnorSponsler opened the pull request:
Adds a new body type EmittingEarthAtm as a derived class from EarthAtm in both body.h and body.cpp that uses TriCubicInterpolator to calculate the differential flux production dependent on coszen, heaght, and energy in the atmosphere.
Modifies nuSQUIDSAtm to allow for a second template that is derived from EarthAtm.
Updates how nuSQUIDSAtm checks for the base class of a template.
Adds functions to nuSQUIDSAtm to allow for the manipulation of the class in the body template.
Adds a new file in examples to that explains the use of nuSQUIDSAtm with the EmittingEarthAtm class.
Adds this example to the configure file so it appears with the old examples in the makefile.
To be added: a new file in data called "atmos_prod" that should contain the differential flux production profile. All data has been calculated, but is currently too large in .dat format to upload.

ConnorSponsler commented on src/body.cpp:
```diff
@@ -29,6 +29,7 @@
 #include <stdexcept>
 #include <type_traits>
 #include <utility>
+#include <fstream>
```
required for reading the datafile in line 966

cnweaver commented on configure:
```diff
@@ -709,6 +709,7 @@ OBJECTS = $(patsubst src/%.cpp,build/%.o,$(SOURCES))
 EXAMPLES := examples/Single_energy/single_energy \
             examples/Multiple_energy/multiple_energy \
             examples/Atm_default/atm_default \
+	    examples/Emitting_Atmosphere/emitting_atmosphere \
```
It looks like this line has gotten a tab character in it, when it should be all spaces for alignment.

cnweaver commented on include/nuSQuIDS/body.h:
```diff
@@ -669,6 +669,9 @@ class EarthAtm: public Body{
     /// \brief Electron fraction at maximum radius.
     double x_ye_max;
+    
+    marray<double, 1> ESpace;
```
I think `ESpace` should be removed everywhere. `EmittingEarthAtm::injected_neutrino_flux` should use `nuSQUIDS::GetERange()` on the passed `nusquids` object to find out what the node energies are at which it should evaluate the flux. This should simplify the code in a number of places (as all functions which exist solely to set/handle it can be removed as well).

cnweaver commented on include/nuSQuIDS/body.h:
```diff
@@ -762,6 +766,123 @@ class EarthAtm: public Body{
+    marray<double,1> heights;
+    /// \brief Atm energies array
+    std::vector<double> atm_energy;
+    marray<double,1> energies;
```
`EmittingEarthAtm` seems to contain a large number of arrays of duplicate data, e.g. `atm_coszen` and `coszens`. At a minimum we should not keep these duplicates, and it isn't clear why most of these things need to be member variables; many look like they could just be local variables within the constructor which takes a file path.

cnweaver commented on src/body.cpp:
```diff
@@ -902,6 +903,145 @@ void EarthAtm::SetAtmosphereHeight(double height){
+      double r = (r2 > 0 ? sqrt(r2) : 0);
+      double rel_r = r / earth_with_atm_radius;
+      double Height = earth_with_atm_radius * (rel_r - radius / earth_with_atm_radius); // here atm_height is a member of EarthAtm in km
+      double H = Height*100000;
```
Unit conversions should be performed using the common unit constants provided by the `params` object.

cnweaver commented on src/body.cpp:
```diff
@@ -902,6 +903,145 @@ void EarthAtm::SetAtmosphereHeight(double height){
+
+      for (unsigned int ei=0; ei < ESpace.extent(0); ei++) {
+        double E = ESpace[ei]/(param.GeV);          
+        if(czNode >= czen_min && czNode<=czen_max && H>=height_min && H<=height_max && E>=energy_min && E<=energy_max){
```
Checks on the zenith angle and height should probably be outside of the loop over energy, to reduce comparisons in the hot loop, and a single, simpler loop could be used in the case that either is out of range to zero out the entire `flux` array.

cnweaver commented on include/nuSQuIDS/body.h:
```diff
@@ -762,6 +766,123 @@ class EarthAtm: public Body{
+    /// \brief indices for the electron and muon flavors
+    /// defaulted to the first and second flavor states
+    int E_nu_index = 0;
+    int Mu_nu_index = 1;
```
I think `E_nu_index` and `Mu_nu_index` should be removed. We should use the usual implicit flavor ordering (nu_e, nu_mu, nu_tau). It would also be good to change all remaining comments an variable names to the standard form (e.g. correct `E_nu` to `nu_e`).

cnweaver commented on include/nuSQuIDS/body.h:
```diff
@@ -762,6 +766,123 @@ class EarthAtm: public Body{
+    void injected_neutrino_flux(marray<double, 3>& flux, const GenericTrack& track, const nuSQUIDS& nusquids) override;
+    
+    //Helpful function for data parsing
+    void unique_sort(std::vector<double>& vec, double& min_val, double& max_val, long unsigned int& number_val);
```
This doesn't appear to need to be either a member function or a part of the public interface; it could just be a free function in the implementation file.

cnweaver reviewed (changes requested):
I think this code is generally on the right track, and will be really great once we tune it up a bit. 

This does need to be extended to handle anti-neutrinos. `EmittingEarthAtm::injected_neutrino_flux` is currently hard-coded to only fill the `rho=0` entries in the `flux` array, but it should be mindful of whether the propagation is neutrinos only, anti-neutrinos only, or both together (which is the most common case people will want to calculate). The input data file should likewise be able to (and probably required to) provide anti-neutrino fluxes. Related to this, the nu_tau flux needs to be handled if the number of flavors is >= 3, even if just by explicitly setting it to zero, to prevent leaving garbage value in the flux array, and likewise any sterile flavors. 

Since it would generally be odd to compute/predict neutrino fluxes without anti-neutrino fluxes, I propose that the text file format should always require them to be present together. It could then have a fixed set of abscissa columns (zenith angle, height, energy), and a variable number of ordinate columns for different neutrino flavors, but with the requirement that the number of ordinate columns should be even. If there are two columns (after the required three), they would be nu_e and nu_e_bar fluxes, if there are four they would be nu_e, nu_e_bar, nu_mu, nu_mu_bar, etc.

cnweaver commented on src/body.cpp:
```diff
@@ -902,6 +903,145 @@ void EarthAtm::SetAtmosphereHeight(double height){
+  marray<double, 3> nuMumatrix{czen_number, height_number, energy_number};
+  
+
+  std::ifstream InputFile(filepath);
```
We shouldn't read the file again; unless I misunderstand we already have all of this data in `atm_prod_model`, so at a minimum we should loop over that instead of repaeting all of the text parsing.

ConnorSponsler commented:
Ok. I believe I've fixed all of these except the removal of the ability to
change the nu_e and nu_mu indices. However, I did not include the ability
to change them in the example file and I added multiple warnings in body.h
telling the user what else they need to do in order to safely use the
function that redefines the indices.  Attached is a google drive link to
the new data file with antineutrino fluxes as well, to replace the old  one
in the same folder.
 PROD_MODEL_MCEQ.dat
<https://drive.google.com/file/d/15iCA9PpLWriVlSVG1TAHmuLRj8VY8a8q/view?usp=drive_web>

On Tue, Nov 14, 2023 at 9:49 AM C. Weaver ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/body.cpp
> <https://github.com/arguelles/nuSQuIDS/pull/41#discussion_r1392715027>:
>
> > +  heights.resize(0,atm_heights.size());
> +  for (unsigned int i=0; i < height_number; i++){
> +    heights[i] = atm_heights[i];
> +  }
> +
> +  unique_sort(atm_energy, energy_min, energy_max, energy_number);
> +  energies.resize(0,atm_energy.size());
> +  for (unsigned int i=0; i < energy_number; i++){
> +    energies[i] = atm_energy[i];
> +  }
> +
> +  marray<double, 3> nuEmatrix{czen_number, height_number, energy_number};
> +  marray<double, 3> nuMumatrix{czen_number, height_number, energy_number};
> +
> +
> +  std::ifstream InputFile(filepath);
>
> We shouldn't read the file again; unless I misunderstand we already have
> all of this data in atm_prod_model, so at a minimum we should loop over
> that instead of repaeting all of the text parsing.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arguelles/nuSQuIDS/pull/41#pullrequestreview-1729925494>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH46WOKPBB6NQYDISRTXTA3YEOAIPAVCNFSM6AAAAAA7B4YA76VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOMRZHEZDKNBZGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
